botocore/utils.py

```diff
@@ -175,9 +176,16 @@ def _get_request(self, url, timeout, num_attempts=1):
             logger.debug("Access to EC2 metadata has been disabled.")
             raise _RetriesExceededError()
 
+        headers = {
+            'User-Agent': 'aws-sdk-botocore/%s ' % botocore_version,
```

Comment:
This would make it inconsistent with what the other SDKs have.  The PR makes the User-Agent used when talking to the metadata service consistent.

Reply:
I think we would prefer to stay consistent with user agent header value that we provide to all of the other AWS API's. We know that the actual user agent value that `botocore` uses today is inconsistent with what the other SDK's use, and we are fine with that inconsistency. So in the end, we just want to avoid having another layer of inconsistency within botocore itself by having some of its requests use a base user agent value that differs from what is associated to the session. Let us know what you think.

Reply:
If you want to use something like `Boto3/1.7.19 Python/2.7.12 Linux/4.4.0-122-generic Botocore/1.10.19` then I think that is OK, however not sure why having something more generic like other SDKs is problematic.  One can easily use a whitelist, but it will be a one off.  Happy to chat more about it.

Reply:
Yeah it would be `Boto3/1.7.19 Python/2.7.12 Linux/4.4.0-122-generic Botocore/1.10.19` assuming `boto3` is being used. To be more specific, it is the `user_agent` value attached to the client and session (depending on if using `boto3` or `botocore` client/session) i.e.
```python
>>> import boto3
>>> c = boto3.client('s3')
>>> c.meta.config.user_agent
'Boto3/1.7.21 Python/2.7.10 Darwin/15.6.0 Botocore/1.10.21'
>>> import botocore.session
>>> s = botocore.session.Session()
>>> s.user_agent()
'Botocore/1.10.21 Python/2.7.10 Darwin/15.6.0'
```
I think along with the consistency argument, the other big problem (and probably more concerning) is that we are breaking the contract for `user_agent` if we are not using the value attached to the client/session when setting the user agent. The `user_agent` is documented more here:
https://github.com/boto/botocore/blob/dbfd6ee157c61222a4090d6191ce2937f12d184f/botocore/session.py#L478-L499
and here:
https://github.com/boto/botocore/blob/dbfd6ee157c61222a4090d6191ce2937f12d184f/botocore/config.py#L31-L32

But the contract essentially implies that whatever value `user_agent` is, the SDK will use that value when injecting the `User-Agent` header. So, the way I see it right now is that we are breaking that contract by using the default header set by `requests`, but by setting it to something other than what is attached the client/session would still be breaking the `user_agent` contract. 

Let us know if that makes sense or you have any thoughts.

Reply:
I think you have a `user_agent` contract with AWS Service APIs for what shows in CloudTrail.  There exists no contract to the meta-data service, however there is an emerging contract across SDKs to use the pattern in this PR.  Just trying to standardize across SDKs and not have one offs to make things easier, especially around automated testing of enforcement.  Having to understand all the different ways to create sessions and if you choose to override the UserAgent your application uses when talking to the AWS APIs makes creating a whitelist for this protection impossible.

Reply:
@willbengtson Can you share over some examples of how this is becoming the new standard contract?  Maybe link to a few PR's of other SDKs exhibiting this preferred contract?